### PR TITLE
Add normals to pcl samples

### PR DESCRIPTION
**Pull request by SatjaSivcev**

### CODE REVIEW

**Comment by runenordmo** [comment]
Can you add the reason why this change is made to the first commit message?
And also how the sample differs from and complements [CaptureHDRVisNormals](https://github.com/zivid/zivid-cpp-samples/blob/master/source/Applications/Basic/Visualization/CaptureHDRVisNormals/CaptureHDRVisNormals.cpp)?

File: source/Applications/Basic/Visualization/CaptureWritePCLVis3D/CaptureWritePCLVis3D.cpp
Comment (runenordmo) [inline source/Applications/Basic/Visualization/CaptureWritePCLVis3D/CaptureWritePCLVis3D.cpp:30]: I think remove the Zivid visualization here, since the sample https://github.com/zivid/zivid-cpp-samples/blob/master/source/Applications/Basic/Visualization/CaptureVis3D/CaptureVis3D.cpp is for that

File: README.md
Comment (runenordmo) [inline README.md:57]: I think update the description to match the `ReadPCLVis3D`, since they correspond.
Maybe `Capture point cloud with/without normals [...]` and `Read point cloud with/without normals [...]`

File: source/Applications/Basic/Visualization/CaptureWritePCLVis3D/CaptureWritePCLVis3D.cpp
Comment (runenordmo) [reply source/Applications/Basic/Visualization/CaptureWritePCLVis3D/CaptureWritePCLVis3D.cpp:30]: And remove corresponding header and linking to Zivid visualizer

Comment (SatjaSivcev) [reply source/Applications/Basic/Visualization/CaptureWritePCLVis3D/CaptureWritePCLVis3D.cpp:30]: done

File: README.md
Comment (SatjaSivcev) [reply README.md:57]: done

**Comment by SatjaSivcev** [comment]
> Can you add the reason why this change is made to the first commit message? And also how the sample differs from and complements [CaptureHDRVisNormals](https://github.com/zivid/zivid-cpp-samples/blob/master/source/Applications/Basic/Visualization/CaptureHDRVisNormals/CaptureHDRVisNormals.cpp)?

done

**Comment by SatjaSivcev** [comment]
Ready for the next round @runenordmo

File: README.md
Comment (runenordmo) [reply README.md]: looks good :+1:

File: source/Applications/Basic/Visualization/CaptureWritePCLVis3D/CaptureWritePCLVis3D.cpp
Comment (runenordmo) [inline source/Applications/Basic/Visualization/CaptureWritePCLVis3D/CaptureWritePCLVis3D.cpp:38]: Would be nice to use an optional for normals here, but cannot use it for c++11 that are in the sample, since optional is c+++ 17

File: source/Applications/Basic/Visualization/CaptureWritePCLVis3D/CaptureWritePCLVis3D.cpp
Comment (runenordmo) [inline source/Applications/Basic/Visualization/CaptureWritePCLVis3D/CaptureWritePCLVis3D.cpp:68]: It's not needed to have to viewports here

Comment (SatjaSivcev) [reply source/Applications/Basic/Visualization/CaptureWritePCLVis3D/CaptureWritePCLVis3D.cpp:68]: My intention was to display the point cloud with color on the left side and the point cloud with color and normals on the right side. Are you saying I should only keep one point cloud with color and normals?

Comment (SatjaSivcev) [reply source/Applications/Basic/Visualization/CaptureWritePCLVis3D/CaptureWritePCLVis3D.cpp:68]: I think it looks better with both:

image

If there is no color in the point cloud, the right side seems like we are only showing normals.

image

Comment (runenordmo) [reply source/Applications/Basic/Visualization/CaptureWritePCLVis3D/CaptureWritePCLVis3D.cpp:68]: I don't think we need both, we already have that in another sample

Comment (runenordmo) [reply source/Applications/Basic/Visualization/CaptureWritePCLVis3D/CaptureWritePCLVis3D.cpp:68]: That's the point of the CaptureHDRVisNormals sample we have already

**Comment by SatjaSivcev** [comment]
Ready for the next round @runenordmo

File: source/Applications/Advanced/MaskPointCloud/MaskPointCloud.cpp
Comment (runenordmo) [inline source/Applications/Advanced/MaskPointCloud/MaskPointCloud.cpp:27]: typo, remove "me"

File: source/Applications/Basic/Visualization/CaptureWritePCLVis3D/CaptureWritePCLVis3D.cpp
Comment (runenordmo) [inline source/Applications/Basic/Visualization/CaptureWritePCLVis3D/CaptureWritePCLVis3D.cpp:116]:
```suggestion
        bool useNormals = argc >= 2 && std::string(argv[1]) == "normals";
```

The ternary operator is redundant here

File: source/Applications/Basic/Visualization/ReadPCLVis3D/ReadPCLVis3D.cpp
Comment (runenordmo) [inline source/Applications/Basic/Visualization/ReadPCLVis3D/ReadPCLVis3D.cpp:61]:
```suggestion
        const bool useNormals = argc >= 2 && std::string(argv[1]) == "normals";
```

Redundant ternary operator. Also add const.

File: source/Applications/Advanced/MaskPointCloud/MaskPointCloud.cpp
Comment (runenordmo) [inline source/Applications/Advanced/MaskPointCloud/MaskPointCloud.cpp:130]: I think this function can continue taking the parameter as a const reference (like before).
No need for ownership semantics / manipulating lifetime.

File: source/Applications/Advanced/ROIBoxViaArucoMarker/ROIBoxViaArucoMarker.cpp
Comment (runenordmo) [inline source/Applications/Advanced/ROIBoxViaArucoMarker/ROIBoxViaArucoMarker.cpp:445]: Consider if this should remain const.

File: source/Applications/Advanced/MaskPointCloud/MaskPointCloud.cpp
Comment (runenordmo) [inline source/Applications/Advanced/MaskPointCloud/MaskPointCloud.cpp:20]: Not sure if we need to replace the `pcl::visualization::CloudViewer` here.
For me it's fine if we show both approaches for visualization in the samples.

Comment (runenordmo) [reply source/Applications/Advanced/MaskPointCloud/MaskPointCloud.cpp:20]: If you disagree, then we should make sure to update all `CloudViewer` occurences in the repo, for instance https://github.com/zivid/zivid-cpp-samples/blob/5d2d0366740d2c2bb725a6689ebd635251b7c713/source/Applications/Advanced/MultiCamera/StitchByTransformation/StitchByTransformation.cpp#L204.

Comment (runenordmo) [reply source/Applications/Advanced/MaskPointCloud/MaskPointCloud.cpp:20]: And then we should also make sure to update all includes of the header
`#include <pcl/visualization/cloud_viewer.h>`
to
`#include <pcl/visualization/pcl_visualizer.h>`

Comment (runenordmo) [reply source/Applications/Advanced/MaskPointCloud/MaskPointCloud.cpp:20]: But as mentioned, I don't mind if we have both ways of visualizing in the repo

Comment (SatjaSivcev) [reply source/Applications/Advanced/MaskPointCloud/MaskPointCloud.cpp:20]: It actually worked better (view point was better) so I want to keep it here. There are two other samples where we use the other one.

File: source/Applications/Basic/Visualization/CaptureWritePCLVis3D/CaptureWritePCLVis3D.cpp
Comment (runenordmo) [inline source/Applications/Basic/Visualization/CaptureWritePCLVis3D/CaptureWritePCLVis3D.cpp:10]: Update to `#include <pcl/visualization/pcl_visualizer.h>`?

File: source/Applications/Basic/Visualization/ReadPCLVis3D/ReadPCLVis3D.cpp
Comment (runenordmo) [inline source/Applications/Basic/Visualization/ReadPCLVis3D/ReadPCLVis3D.cpp:10]: Should probably update to `#include <pcl/visualization/pcl_visualizer.h>`

File: source/Applications/Advanced/MaskPointCloud/MaskPointCloud.cpp
Comment (runenordmo) [inline source/Applications/Advanced/MaskPointCloud/MaskPointCloud.cpp:11]: Update to `#include <pcl/visualization/pcl_visualizer.h>` if you decide to use that visualizer

File: source/Applications/Advanced/MaskPointCloud/MaskPointCloud.cpp
Comment (SatjaSivcev) [reply source/Applications/Advanced/MaskPointCloud/MaskPointCloud.cpp:20]: I added the correct includes for all PCL samples

File: source/Applications/Basic/Visualization/ReadPCLVis3D/ReadPCLVis3D.cpp
Comment (runenordmo) [inline source/Applications/Basic/Visualization/ReadPCLVis3D/ReadPCLVis3D.cpp:68]: I don't think this is intended to be const in the way that you write the sample?
You load pcd file data into as the data of the pointer just below?

File: source/Applications/Basic/Visualization/ReadPCLVis3D/ReadPCLVis3D.cpp
Comment (runenordmo) [inline source/Applications/Basic/Visualization/ReadPCLVis3D/ReadPCLVis3D.cpp:78]: Not intended to be const?

**Comment by runenordmo** [comment]
Looks quite good now.
If you can resolve the rest of the threads, and squash + rebase, I can just have a short look before approving :+1:

File: source/Applications/Basic/Visualization/CaptureWritePCLVis3D/CaptureWritePCLVis3D.cpp
Comment (runenordmo) [inline source/Applications/Basic/Visualization/CaptureWritePCLVis3D/CaptureWritePCLVis3D.cpp]: Here you changed back to the non `boost::shared_ptr` version?

File: source/Applications/Basic/Visualization/ReadPCLVis3D/ReadPCLVis3D.cpp
Comment (runenordmo) [inline source/Applications/Basic/Visualization/ReadPCLVis3D/ReadPCLVis3D.cpp:63]: Btw, have you updated the sample pcd file to contain normals?
Else, I guess you cannot do `pcl::io::loadPCDFile<pcl::PointXYZRGBNormal>(pointCloudFile, *pointCloudWithNormalsPCL);`?

File: source/Applications/Basic/Visualization/CaptureWritePCLVis3D/CaptureWritePCLVis3D.cpp
Comment (runenordmo) [reply source/Applications/Basic/Visualization/CaptureWritePCLVis3D/CaptureWritePCLVis3D.cpp]: Sorry, I see that's updated in the second commit

File: source/Applications/Basic/Visualization/CaptureWritePCLVis3D/CaptureWritePCLVis3D.cpp
Comment (runenordmo) [reply source/Applications/Basic/Visualization/CaptureWritePCLVis3D/CaptureWritePCLVis3D.cpp]: Sorry, I see that's updated in the second commit

**Comment by runenordmo** [comment]
Two unresolved comments.
Just ping me to approve when you resolve those.

File: source/Applications/Advanced/MaskPointCloud/MaskPointCloud.cpp
Comment (SatjaSivcev) [reply source/Applications/Advanced/MaskPointCloud/MaskPointCloud.cpp:131]: Then I have to call it like `visualizeDepthMap(*maskedPointCloudPCL);`

Comment (SatjaSivcev) [reply source/Applications/Advanced/MaskPointCloud/MaskPointCloud.cpp:131]: And that makes `main `look like this:

```
        std::cout << "Converting to PCL point cloud" << std::endl;
        const auto pointCloudPCL = convertToPCLPointCloud(pointCloud);

        std::cout << "Displaying point cloud before masking" << std::endl;
        visualizePointCloud(pointCloudPCL);

        std::cout << "Displaying depth map before masking" << std::endl;
        visualizeDepthMap(*pointCloudPCL);

        std::cout << "Masking point cloud" << std::endl;
        const boost::shared_ptr<pcl::PointCloud<pcl::PointXYZRGB>> maskedPointCloudPCL =
            maskPointCloud(pointCloud, mask);

        std::cout << "Displaying point cloud after masking" << std::endl;
        visualizePointCloud(maskedPointCloudPCL);

        std::cout << "Displaying depth map after masking" << std::endl;
        visualizeDepthMap(*maskedPointCloudPCL);
```

Comment (SatjaSivcev) [reply source/Applications/Advanced/MaskPointCloud/MaskPointCloud.cpp:131]: Do you think that is ok?

File: source/Applications/Basic/Visualization/ReadPCLVis3D/ReadPCLVis3D.cpp
Comment (SatjaSivcev) [reply source/Applications/Basic/Visualization/ReadPCLVis3D/ReadPCLVis3D.cpp:63]: I haven't but that is the first thing to do after this gets merged.

File: source/Applications/Advanced/MaskPointCloud/MaskPointCloud.cpp
Comment (runenordmo) [reply source/Applications/Advanced/MaskPointCloud/MaskPointCloud.cpp]: Looks good to me 👍

**Review by runenordmo — APPROVED** [review]

**Pull request merged**
